Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1330 @@
+double = label "double" $ lexeme $ read <$> do
+  left <- some numberChar
+  rightM <- optional $ do
+    char '.'
+    some numberChar
+  char' 'f'  -- Like char, but case-insensitive
+  pure $ case rightM of
+    Nothing -> left
+    Just right -> left ++ "." ++ right
+```
+
+Now, we add it to `atom`: `SDouble <$> double`. Depending whether you added `double` before or after `integer`, the result may be surprising:
+
+<table>
+  <tr>
+    <td>Parser \ Position</td>
+    <td>`double` before `integer`</td>
+    <td>`double` after `integer`</td>
+  </tr>
+  <tr>
+    <td>
+      <pre lang="hs">parseTest atom "42"</pre>
+    </td>
+    <td>
+      <pre>
+1:3:
+  |
+1 | 42
+  |   ^
+unexpected end of input
+expecting '.', 'F', 'f', or numeric character
+      </pre>
+    </td>
+    <td>
+      <pre lang="hs">SDouble 42.0</pre>
+    </td>
+  </tr>
+  <tr>
+    <td>
+      <pre lang="hs">parseTest atom "42f"</pre>
+    </td>
+    <td>
+      <pre lang="hs">SInteger 42</pre>
+    </td>
+    <td>
+      <pre lang="hs">SInteger 42</pre>
+    </td>
+  </tr>
+</table>
+
+So it only got one case correctly. To fix it, one solution would be to use `try`, which allows Megaparsec to do as much backtracking as it wants.
+
+```hs
+-- Note: Other cases omitted for brevity.
+atom :: Parser SExp
+atom = choice
+  [ SDouble <$> try double
+  , SInteger <$> integer
+  ]
+```
+
+And it should fix the problem. However, this is not ideal in terms of performance, as it will cause Megaparsec to consume the entirety of the `double`, to later find that there is no `f`, backtrack, and finally try again with an `integer`.
+
+In such a situation this is trivial, but care must be taken in more complex examples, for example, what if we wanted to have `(foo bar)` as well as tuples such as `(foo bar, baz quz)`? The S-expression could be very long and lead to exponential complexity in the worst-case scenario. Recently, a very similar problem [was fixed](https://gitlab.com/morley-framework/morley/-/merge_requests/971) in one of our projects.
+
+In this situation, the best thing to do is to _refactor the grammars to share common parts_. The code above could be written without `try` as such:
+
+```hs
+numeric :: Parser SExp
+numeric = label "number" $ lexeme $ do
+  -- Left side before . or f
+  left <- some numberChar
+
+  -- .0f or f
+  rightM <- optional $ choice
+    [ do
+        char '.'
+        right <- some numberChar
+        char' 'f'
+        pure $ left ++ "." ++ right
+    , do
+        char' 'f'
+        pure left
+    ]
+
+  -- If we had the right side, it's a double, otherwise an integer.
+  pure $ case rightM of
+    Nothing -> SInteger $ read left
+    Just right -> SDouble $ read right
+```
+
+And in `atom`, we simply use `numeric` instead of the two previous parsers.
+
+## Rewriting with Megaparsec's lexer
+
+A lot of the things could be written using the lexer module from Megaparsec. We present the simplified versions below for functions that can be improved.
```

Comment:
I would organize it a little bit, at least the 'improvements" should be below the code. 

Because you said "we present the simplified versions below" and I was reading lines like "where's the code, Lebowski?"

Reply:
Good point, reorganized and reworded this section a bit.